### PR TITLE
move hypervisor vendor id code into cpuid

### PR DESCRIPTION
**Pull request by brianmcgee**
TODO

- [x] Get https://github.com/klauspost/cpuid/pull/147 accepted upstream
- [x] Switch to upstream

### CODE REVIEW

File: pkg/virt/detect_vm_cpu_id.go [hunk 1]
```diff
@@ -8,44 +8,29 @@ import (
...
-	"Microsoft Hv": TypeMicrosoft,
-	/* https://wiki.freebsd.org/bhyve */
-	"bhyve bhyve ": TypeBhyve,
-	"QNXQVMBSQG":   TypeQnx,
-	/* https://wiki.freebsd.org/bhyve */
-	"ACRNACRNACRN": TypeAcrn,
-	/* https://www.lockheedmartin.com/en-us/products/Hardened-Security-for-Intel-Processors.html */
-	"SRESRESRESRE": TypeSre,
-	"Apple VZ":     TypeApple,
+var vendorMapping = map[cpuid.Vendor]Type{
```
Comment (Mic92) [inline pkg/virt/detect_vm_cpu_id.go:11]: cpuid is an x86 instruction, won't we limit ourself to x86 this way?

Comment (Mic92) [reply pkg/virt/detect_vm_cpu_id.go:11]: What does systemd use here btw?
Looks like the library also works on arm64.

Comment (Mic92) [reply pkg/virt/detect_vm_cpu_id.go:11]: At least qemu can also change cpuids, so I am not sure that this detection is always robust. Systemd's code uses procfs instead: https://github.com/systemd/systemd/blob/main/src/basic/virt.c#L159

Comment (Mic92) [reply pkg/virt/detect_vm_cpu_id.go:11]: Looks more portable across different architectures and doesn't require any third-party libraries.

Comment (Mic92) [reply pkg/virt/detect_vm_cpu_id.go:11]:
```
package main

import (
    "bufio"
    "log"
    "os"
    "strings"
)

// Define Virtualization types
type Virtualization int

const (
    VIRTUALIZATION_NONE Virtualization = iota
    VIRTUALIZATION_KVM
    VIRTUALIZATION_AMAZON
    VIRTUALIZATION_QEMU
    VIRTUALIZATION_VMWARE
    VIRTUALIZATION_ORACLE
    VIRTUALIZATION_XEN
    VIRTUALIZATION_BOCHS
    VIRTUALIZATION_PARALLELS
    VIRTUALIZATION_BHYVE
    VIRTUALIZATION_MICROSOFT
    VIRTUALIZATION_APPLE
    VIRTUALIZATION_GOOGLE
)

// Struct for vendor and virtualization ID
type DmiVendor struct {
    vendor string
    id     Virtualization
}

// List of DMI vendor paths
var dmiVendors = []string{
    "/sys/class/dmi/id/product_name",
    "/sys/class/dmi/id/sys_vendor",
    "/sys/class/dmi/id/board_vendor",
    "/sys/class/dmi/id/bios_vendor",
    "/sys/class/dmi/id/product_version",
}

// Table of known DMI vendors and their virtualization IDs
var dmiVendorTable = []DmiVendor{
    {"KVM", VIRTUALIZATION_KVM},
    {"OpenStack", VIRTUALIZATION_KVM},
    {"KubeVirt", VIRTUALIZATION_KVM},
    {"Amazon EC2", VIRTUALIZATION_AMAZON},
    {"QEMU", VIRTUALIZATION_QEMU},
    {"VMware", VIRTUALIZATION_VMWARE},
    {"VMW", VIRTUALIZATION_VMWARE},
    {"innotek GmbH", VIRTUALIZATION_ORACLE},
    {"VirtualBox", VIRTUALIZATION_ORACLE},
    {"Oracle Corporation", VIRTUALIZATION_ORACLE},
    {"Xen", VIRTUALIZATION_XEN},
    {"Bochs", VIRTUALIZATION_BOCHS},
    {"Parallels", VIRTUALIZATION_PARALLELS},
    {"BHYVE", VIRTUALIZATION_BHYVE},
    {"Hyper-V", VIRTUALIZATION_MICROSOFT},
    {"Apple Virtualization", VIRTUALIZATION_APPLE},
    {"Google Compute Engine", VIRTUALIZATION_GOOGLE},
}

func readOneLineFile(path string) (string, error) {
    file, err := os.Open(path)
    if err != nil {
        return "", err
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    if scanner.Scan() {
        return scanner.Text(), nil
    }

    return "", scanner.Err()
}

func detectVmDmiVendor() Virtualization {
    for _, vendorPath := range dmiVendors {
        s, err := readOneLineFile(vendorPath)
        if err != nil {
            if os.IsNotExist(err) {
                continue
            }
            log.Printf("Error reading file %s: %v", vendorPath, err)
            return VIRTUALIZATION_NONE
        }

        for _, entry := range dmiVendorTable {
            if strings.HasPrefix(s, entry.vendor) {
                log.Printf("Virtualization %s found in DMI (%s)", s, vendorPath)
                return entry.id
            }
        }
    }
    log.Println("No virtualization found in DMI vendor table.")
    return VIRTUALIZATION_NONE
}

func main() {
    vmType := detectVmDmiVendor()
    log.Printf("Detected Virtualization: %v", vmType)
}

```

Comment (Mic92) [reply pkg/virt/detect_vm_cpu_id.go:11]: cc @nikstur anything I am missing here?

Comment (nikstur) [reply pkg/virt/detect_vm_cpu_id.go:11]: Looks good.

**Pull request merged**